op-tests/test-bls-ops.txt

```diff
@@ -0,0 +1,209 @@
+bls_g2_negate 0 => FAIL
+bls_g2_negate ( 1 2 3 ) => FAIL
+
+; bls map to g1
+; expects (seed [DST_str])
+bls_map_to_g1 "abcdef0123456789" => 0x88e7302bf1fa8fcdecfb96f6b81475c3564d3bcaf552ccb338b1c48b9ba18ab7195c5067fe94fb216478188c0a3bef4a | 619144
+bls_map_to_g1 "abcdef0123456789" "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_NUL_" => 0x8dd8e3a9197ddefdc25dde980d219004d6aa130d1af9b1808f8b2b004ae94484ac62a08a739ec7843388019a79c437b0 | 619144
+bls_map_to_g1 "abcdef0123456789" "BLS_SIG_BLS12381G1_XMD:SHA-256_SSWU_RO_AUG_" => 0x88e7302bf1fa8fcdecfb96f6b81475c3564d3bcaf552ccb338b1c48b9ba18ab7195c5067fe94fb216478188c0a3bef4a | 619144
+; invalid number of arguments
+bls_map_to_g1 => FAIL
+bls_map_to_g1 1337 42 10 => FAIL
+; invalid arguments
+bls_map_to_g1 ( "hello" ) => FAIL
+bls_map_to_g1 ( "foo" "bar" ) => FAIL
+bls_map_to_g1 (()) => FAIL
+
+; bls map to g2
+; expects (seed [DST_str])
+
+bls_map_to_g2 "abcdef0123456789" => 0x92596412844e12c4733b5a6bfc5727cde4c20b345665d2de99de163266f3ba6a944c6c0fdd9d9fe57b9a4acb769bf3780456f8aab4cd41a70836dba57a5278a85fbd18eb96a2b56cfbda853186c9d190c43e63bc3e6a181aed692e97bbdb1944 | 3388740
+bls_map_to_g2 "abcdef0123456789" "BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_NUL_" => 0x8ee1ff66094b8975401c86ad424076d97fed9c2025db5f9dfde6ed455c7bff34b55e96379c1f9ee3c173633587f425e50aed3e807c6c7cd7bed35d40542eee99891955b2ea5321ebde37172e2c01155138494c2d725b03c02765828679bf011e | 3388740
+bls_map_to_g2 "abcdef0123456789" "BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_AUG_" => 0x92596412844e12c4733b5a6bfc5727cde4c20b345665d2de99de163266f3ba6a944c6c0fdd9d9fe57b9a4acb769bf3780456f8aab4cd41a70836dba57a5278a85fbd18eb96a2b56cfbda853186c9d190c43e63bc3e6a181aed692e97bbdb1944 | 3388740
+; invalid number of arguments
+bls_map_to_g2 => FAIL
+bls_map_to_g2 1337 42 10 => FAIL
+; invalid arguments
+bls_map_to_g2 ( "hello" ) => FAIL
+bls_map_to_g2 ( "foo" "bar" ) => FAIL
+bls_map_to_g2 (()) => FAIL
+
+
+; bls_pairing_identity
+; expects ((G1 . G2) (G1 . G2) ... )
+
+bls_pairing_identity ((0x8b202593319bce41b090f3309986de59861ab1e2ff32aef871d83f9aac232c7253c01f1f649c6f69879c441286319de4 . 0x942adad4dbeadcfd75aaa11940a5e5e16a8d8e91742029a3944610635ccc0572eceeb1c89d8a0e904c5d30b9497e700312dee7b833535effef24953dbf8f8aa770e2f1a8e01d3b6f6844e01a635ed95664babe9d62a2572651d0258461c8ba00) (0xb7f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb . 0x80c37921e62092ef55f85f9eccb21bd80cfaafc0bce9cbdd6999b1a8cabadc8f23720f0261efafaf53cbcc74580b9432007b66d824668900a94934f184bc41bf9ccf9ec141c6f7da610aa7296cd0a181ae8fe176b607aa4c367f15ee0cb985d7) ) => 1 | 14029963
```

Comment:
This is one of the things I'm interested in your feedback on. Chialisp *does* support the `.`-syntax, and it just means you explicitly specify the right-hand side of a cons-box. In this case, the arguments are pairs, and this is a compact representation of pairs. If it had been triples, it wouldn't have worked.

Another odd aspect of using `.` is that you can write a list that's not null-terminated, like this:
```
(1 2 . 3)
```

Which means: `(1 . (2 . 3))`

Cameron's original patch encoded these pairs as lists of 2 items, which made the code a bit more complex, because there are more preconditions to check for, and the internal structure a bit less efficient, since you need more cons-boxes for it. That would looks like this:

```
((1 . (2 . NIL)))
```

instead of:

```
(1 . 2)
```

Reply:
Yeah I think cons boxes as an ecoding for the arguments are fine.  You can use the dot syntax more than I thought however, it's a little fuzzy when it works and when it doesn't.

```
(mod () 
  (defun add (list_to_add)
    (+ . list_to_add)
  )

  (add (list 1 2))
)
```
^ It doesn't like this for example.  My opinion is that if `(bls_pairing_identity (c blah blah) (c blah blah))` works then I'm fine with it, regardless of whether or not `(bls_pairing_identity (blah . blah) (blah . blah))` also works.

Reply:
yeah, that highlights another quirk (perhaps). `bls_pairing_identity` takes a *single* argument, which is a list of pairs. so it would be:

```
(bls_pairing_identity (list (blah . blah) (blah . blah)))
```

(using `c` as you suggest would also work).

As I was documenting this today, I realized that I could make this operator just take a variable number of arguments instead, where each argument is a pair. Perhaps that would be more natural. However, it would make it differ more from its sibling function `bls_verify` which takes exactly *two* arguments, a list of pairs and a signature. So, `bls_verify` would still need to take a list of pairs, rather than taking a variable number of arguments.

Reply:
I suppose if the signature comes first, it could still be a variable number of arguments *after* the signature...

Reply:
I'm updating the signatures for `bls_pairing_identity` and `bls_verify` to both take a variable number of arguments